### ISSUE TITLE
Cannot quit program whilst a thread is waiting on a semaphore.

### ISSUE THREAD

**Issue by JonasMht** [godotengine/godot#47758]
**Godot version:** 3.3-rc8

**OS/device including version:** Ubuntu 20.04.2 LTS

**Issue description:**
Played scene won't terminate on get_tree().quit() and will freeze whilst a thread is still running and waiting on a semaphore.
Expected behavior: program termination (as in 3.2.3).

**Steps to reproduce:**
Step1: Create a main scene with a gdscript script attached.
Step2: Add the following code to the script (after include):
```gdscript
var thread = Thread.new()
var sem = Semaphore.new()


func thread_func(_u) -> void:
	while true:
		sem.wait()
		print("Hello")

func _ready():
	sem.post()
	thread.start(self, "thread_func", 0)
	get_tree().quit()
```
Step3 : Run<br>

**Minimal reproduction project:**

[Thread_and_Semaphore_Issue.zip](https://github.com/godotengine/godot/files/6288390/Thread_and_Semaphore_Issue.zip)

**Comment by Calinou** [comment]
This is likely a regression from https://github.com/godotengine/godot/pull/45618. Can you try reproducing this bug in earlier RCs to make sure?

**Comment by akien-mga** [comment]
I don't think this is a regression, this code is invalid. You need to join the thread and also an exit condition in the while loop. See https://docs.godotengine.org/en/3.3/tutorials/threads/using_multiple_threads.html

The fact that it could exit in 3.2.3 could be considered a bug, as it left your thread hanging in an invalid state.

**Comment by JonasMht** [comment]
> I don't think this is a regression, this code is invalid. You need to join the thread and also an exit condition in the while loop. See https://docs.godotengine.org/en/3.3/tutorials/threads/using_multiple_threads.html
> 
> The fact that it could exit in 3.2.3 could be considered a bug, as it left your thread hanging in an invalid state.

Thanks for the feedback.

This invalid code can be found in the documentation:
https://docs.godotengine.org/en/stable/tutorials/io/background_loading.html
The code provided is the following:
https://docs.godotengine.org/en/stable/_downloads/bc54982daa1be2e5b495d74f01a098f1/resource_queue.gd

```gdscript
func thread_func(_u):
	while true:
		thread_process()
```

As you can see the thread is never ended.

**Comment by JonasMht** [comment]
> This is likely a regression from #45618. Can you try reproducing this bug in earlier RCs to make sure?

After testing I found that the change occurred between RC2 and RC3 with RC2 exiting and RC3 freezing.
